govcd/vm.go

```diff
@@ -850,3 +851,186 @@ func (vm *VM) SetGuestCustomizationSection(guestCustomizationSection *types.Gues
 
 	return vm.GetGuestCustomizationSection()
 }
+
+// AddInternalDisk creates disk type *types.DiskSettings to the VM.
+// Returns new disk ID and error.
+// Runs synchronously, VM is ready for another operation after this function returns.
+func (vm *VM) AddInternalDisk(diskData *types.DiskSettings) (string, error) {
+	if vm.VM.HREF == "" {
+		return "", fmt.Errorf("cannot add internal disks - VM HREF is unset")
+	}
+
+	err := vm.validateInternalDiskInput(diskData)
+	if err != nil {
+		return "", err
+	}
+
+	diskSettings := vm.VM.VmSpecSection.DiskSection.DiskSettings
+	if diskSettings == nil {
+		diskSettings = []*types.DiskSettings{}
+	}
+
+	diskSettings = append(diskSettings, diskData)
+	vmSpecSection := vm.VM.VmSpecSection
+	vmSpecSection.DiskSection.DiskSettings = diskSettings
+
+	vmSpecSection, err = vm.UpdateInternalDisks(vmSpecSection)
+	if err != nil {
+		return "", err
+	}
+
+	for _, diskSetting := range vmSpecSection.DiskSection.DiskSettings {
+		if diskSetting.AdapterType == diskData.AdapterType &&
+			diskSetting.BusNumber == diskData.BusNumber &&
+			diskSetting.UnitNumber == diskData.UnitNumber {
+			return diskSetting.DiskId, nil
+		}
+	}
+
+	return "", fmt.Errorf("created disk wasn't in list of returned VM internal disks")
+}
+
+func (vm *VM) validateInternalDiskInput(diskData *types.DiskSettings) error {
+	if diskData.AdapterType == "" {
+		return fmt.Errorf("disk settings missing required field: adapter type")
+	}
+
+	if diskData.BusNumber < 0 {
+		return fmt.Errorf("disk settings bus number has to be 0 or higher")
+	}
+
+	if diskData.UnitNumber < 0 {
+		return fmt.Errorf("disk settings unit number has to be 0 or higher")
+	}
+
+	if diskData.SizeMb < int64(0) {
+		return fmt.Errorf("disk settings size MB has to be 0 or higher")
+	}
+
+	if diskData.Iops != nil && *diskData.Iops < int64(0) {
+		return fmt.Errorf("disk settings iops has to be 0 or higher")
+	}
+
+	if diskData.ThinProvisioned != nil {
```

Comment:
Shouldn't it be the opposite? 
```suggestion
	if diskData.ThinProvisioned == nil {
```

Same applies to below `StorageProfile` field


```shell
vm_test.go:1065:
    check.Assert(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"disk settings missing required field: thin provisioned"} ("disk settings missing required field: thin provisioned")
```

Reply:
ty. Changed